First draft of trends capability

oneiros opened the pull request:
This is modeled closely after what mastodon currently displays on the "explore" page.

This does not mean that we think there are no other use cases, it is just what we are most familiar with.

As usual, any feedback is very welcome, but I would especially like to hear from other implementers if this would be useful to them and what is possibly missing.

oneiros commented on discovery/trends/v0.1/trends.md:
```diff
@@ -0,0 +1,355 @@
+  server uses several FASP to query for trends that all run the same
+  software, it MUST be possible to merge results according to rank to
+  get the correct order.
+* `distribution`: An object with one key per hour requested. The keys
+  are consecutive integers each representing one hour. The first hour
+  requested in chronological order is represented by the number `1`. The
+  value of each key is an integer between 0 and 100 representing how
+  much this result was trending in that hour. This information can
+  optionally be used by fediverse servers to display small diagrams
+  ("sparklines") or arrows marking upward and downward trajectories.
```
This is inspired by the little "sparklines" graph we display for trending hashtags:

![image](https://github.com/user-attachments/assets/a2d2d03b-52c4-4a2a-8f3d-0da4b311f0d0)

ThisIsMissEm commented on discovery/trends/v0.1/trends.md:
````diff
@@ -0,0 +1,355 @@
+hours and to no more than 10 results:
+
+```http
+GET /trends/v0/content?withinLastHours=2&maxCount=10
````
You'd possibly be better using unix epoch here, because servers may have clock skew

oneiros replied:
I am not sure if clock skew should be much of an issue here (unless it is very severe and/or permanent). Worst case the trends are a tad bit older than anticipated.

Some of the reasons for choosing an integer number of hours are:

* The API is explicitly *not* meant to query for historic trend data.
* This effectively sets the resolution of trend data to full hours. This can have a huge impact on the volume of data FASP have to store.

nikclayton commented:
# On the different sections in the doc

## Representation of Content, Hashtags, and Links

Please **do not** case fold hashtags.

First, eventually a user agent will need to display the hashtag. It is an accessibility issue for people using screenreaders to case fold a #HashtagLikeThis (which can be read out as three separate words) to #hashtaglikethis (which will not be).

Second, #Superbowl and #SuperbOwl are two different hashtags.

Third, you cannot reliably case fold in some languages and get the hashtag that people are actually using. See, for example, https://github.com/python/cpython/blob/fff3c28052e6b0750d6218e00acacd2fded4991a/Lib/logging/handlers.py#L802

## Common Request Options

`withinLastHours`

I don't recommend putting the maximum (336) in the spec. Allow clients to request arbitrarily large numbers, and in the response include a property that describes how many hours have been returned (or, just let clients calculate it from the max of the keys in `distribution`, although those are strings which will need to be coerced back to ints.

Optionally, either also include a property that specifies the maximum value this FASP server supports, or provide a specific API call that allows clients to query these values.

`language` - ISO 639 defines four sets of language identifiers (https://www.iso.org/iso-639-language-code). Please be very explicit about which set you mean.

For hashtags and links I assume the language filter is applied to the language of the post that contained the link or the hashtag. If so, please be explicit about that.

## Common Response Attributes

`rank` - suppose a client does consult multiple FASP to get the results. How is the client supposed to merge results where the rank for two or more items are identical? What algorithm is it supposed to use the break the tie?

`distribution` - says this has one key per hour requested. Suppose I request `336`, but the server can only provide the last 10 hours (maybe the server is very new, and doesn't have 336 hours worth of data yet).

What happens? Does the server ignore the `336` and return 10 values? Or does it return 336 values but most of them are zero (or some other sentinel value to indicate the server does not have this data)? Or does it return an error (note: there's no obvious HTTP 5xx error code to return).

How can `distribution` be compared within or across results?

For example, in this response snippet:

```
    {
      "uri": "https://fedi1.example.com/status/23",
      "rank": 100,
      "distribution": {
          "1": 80,
          "2": 90,
          "3": 100
      }
    },
    {
      "uri": "https://fedi3.example.com/posts/17",
      "rank": 74,
      "distribution": {
          "1": 70,
          "2": 90,
          "3": 40 
      }
    },
```

Does this mean that, at hour 2, both those posts had the same distribution? Or is the distribution's value only relative to the other values for that item?

What about across types? If a content response, links response, and hashtags response all have the same value for an item at the same hour does that mean they are equally "trending"?

The same question applies to the `rank` value -- if it's the same across two different types (link and hashtag, say) does that mean they're equally trending?

Alternatively, why not just return the absolute numbers here, and let the client figure out the best way to represent this data to the user? See https://mastodon.social/@nikclayton/110520209653142761 for an example.

Shouldn't `language` be a common response attribute? Especially if the client has omitted it in the request, it will need to know the language of item in the response array (it might want to do its own filtering or grouping).


## Requesting trending content

The spec says "These objects MUST be sorted by `rank` in descending order". Why? This is a client / data presentation concern.

In practice clients are not going to make this query, then confirm the data is sorted by rank and throw it away if it's not. It's simpler just to tell clients they will need to sort the results (see also the comment later about data freshness for other ways clients may wish to sort results).

Is there a better name for this than "content"?

# Other notes

## Data freshness

The responses do not indicate how fresh the data is. Clients may legitimately want to show the user when the trends were last updated, please consider including this in the response.

## Cache control headers

Related to that, you might want to include recommendations / best practices for the different cache-control headers the server might want to include in the response.

## Versioning responses

Please commit to bumping the API endpoint version any time the response format changes in a backwards compatible way.

By comparison, Mastodon does not do this, and it makes life more difficult for client authors.

## `uri` vs `url`

The `contents` API uses a property called `uri`, the `links` one is called `url`. They should probably be the same.

## One API call to get all three?

Is there any reason not to also support an API call that allows a client to request data about all three types in a single call? It's more efficient than making three API calls, parsing three responses, etc.

`GET /trends/v0?kind=links,hashtags` for example, to get a response with both the `links` and `hashtags` properties?

nikclayton commented:
You mentioned this is modeled on Mastodon's trends implementation.

It uses a different API endpoint, but Mastodon also has the concept of a trending account. It's part of the suggestions API (https://docs.joinmastodon.org/entities/Suggestion/) where some of the reasons for recommending an account fall under "trending";

```
most_followed = This account has many active local followers
most_interactions = This account had many reblogs and favourites within the last 30 days
```

You might also want to consider an account as trending if it's number of followers has grown recently relative to other accounts.

There are other criteria you might want to use when filtering trending posts. For example, allow the client to fetch a list of trending posts (`content`) limited to just posts that have video attachments to find trending videos. Or only image attachments. Or polls.

Also provide finer control over what is considered to be trending. The document says a trending post might be trending because of replies, favourites, or boosts. But a user might be specifically interested in "Which posts are most favourited" or "Which threads are most active", so allow them to filter by those criteria.

oneiros commented:
Hi Nik, thanks a lot for your feedback. I will try to address most of it below:

> Please **do not** case fold hashtags.

I hate to inform you, but Mastodon already does this. But this is not about Mastodon, but a provider that can be queried for a couple of hashtags that are relevant right now.

One of our goals here is to minimize the risk of all the top results being a variation of the same thing. We all know that if something is discussed a lot, you will see all kinds of (mis)spellings of hashtags.

The downside is of course that two unrelated hashtags could (for a short while) be considered the same. Not ideal, but I believe it is worth making _some_ trade-off here. What exactly this trade-off is, is not part of the specification. There are no hard `MUST`s here. But implementers should be aware of this issue.

> `withinLastHours`
> 
> I don't recommend putting the maximum (336) in the spec.

I am not convinced of this either, especially since I am not sure how useful two weeks really are.

Putting a limit on this does have two separate purposes, though:

1. This imposes a limit on how much data FASP need to store. Discovery FASP will be resource-hungry, and if we can alleviate this even a little bit, I think it might be worth it.
2. We are very conscious of specific privacy concerns as well as more unspecific aversion against any kind of unnecessary accumulation of data. That is why this capability is only about _current_ trends and why I believe setting a limit on data-retention might make sense.

> `language` - ISO 639 defines four sets of language identifiers (https://www.iso.org/iso-639-language-code). Please be very explicit about which set you mean.

You caught me. I had hoped to get away with this :slightly_smiling_face: 

I was thinking about what we do in Mastodon, which is basically this here: https://github.com/mastodon/mastodon/blob/4a2813158d16914450b6bc4c00e1b3e7edfa26ba/app/helpers/languages_helper.rb

This amounts to using two-letter variants where available and three-letter ones if no two-letter one exists. I am not sure how to succinctly describe this in the spec.

But I did think of another issue here: There are cases like Portuguese, where people in Portugal might not be interested in what is trending right now in Brazil and vice versa. This is less of a problem for e.g. German, which is mostly spoken in neighboring countries. For some reason I have never heard this concern from English speaking countries either.

I am not sure how to solve that.

> For hashtags and links I assume the language filter is applied to the language of the post that contained the link or the hashtag. If so, please be explicit about that.

Good point.

> `rank` - suppose a client does consult multiple FASP to get the results. How is the client supposed to merge results where the rank for two or more items are identical? What algorithm is it supposed to use the break the tie?

I am not sure this matters much, but if you have a preference, I am happy to put it in for completeness sake.

> `distribution` - says this has one key per hour requested. Suppose I request `336`, but the server can only provide the last 10 hours (maybe the server is very new, and doesn't have 336 hours worth of data yet).
> 
> What happens? Does the server ignore the `336` and return 10 values? Or does it return 336 values but most of them are zero (or some other sentinel value to indicate the server does not have this data)? Or does it return an error (note: there's no obvious HTTP 5xx error code to return).

Thanks for spotting this edge case. I do not think this warrants an error response, but I am a bit torn between returning zeros (easier for the fediverse server to handle) and only returning the data that is actually available (more correct).

Would you have a preference?

> How can `distribution` be compared within or across results?

Please note that this is only there, so we can still render those small "sparklines" graph. I am not currently aware of any other use case for this. If you think there is and that this warrants further specification, I am curious to hear about it.

Also, if people believe this is a strange and underspecified thing to have here just so that one platform can have nice tiny graphs, I can also imagine removing this "distribution" object for now.

> The same question applies to the `rank` value -- if it's the same across two different types (link and hashtag, say) does that mean they're equally trending?

No, I believe FASP should have the freedom to calculate trends very differently for the different entities.

> Alternatively, why not just return the absolute numbers here, and let the client figure out the best way to represent this data to the user? See https://mastodon.social/@nikclayton/110520209653142761 for an example.

There is no definition what those absolute numbers would be. We deliberately leave it up to FASP to decide which numbers to take into account and how to calculate trends from them.

> Shouldn't `language` be a common response attribute? Especially if the client has omitted it in the request, it will need to know the language of item in the response array (it might want to do its own filtering or grouping).

Maybe :slightly_smiling_face: 

> The spec says "These objects MUST be sorted by `rank` in descending order". Why? This is a client / data presentation concern.

If I can choose between one server having to sort results vs. every client having to do it, I tend to choose the former. I also think there is no harm in doing so.

> Is there a better name for this than "content"?

Very likely yes. But we have not found it yet. There were drafts of the data sharing capability where I tried to use concrete AS object types to be very precise. But we decided that we cannot afford this level of precision at this stage, because we do not yet know all the use cases, which other implementers are interested in using this and so on.

So for now we settled on "content" even though it is a bit unsatisfying. I hope we can replace this before the specs reach 1.0.

> The responses do not indicate how fresh the data is. Clients may legitimately want to show the user when the trends were last updated, please consider including this in the response.

What do you mean exactly? As one can only request trends up to the current point in time, data is always expected to be fresh.

> Related to that, you might want to include recommendations / best practices for the different cache-control headers the server might want to include in the response.

I do not think we can cache anything at the HTTP level here.

> Please commit to bumping the API endpoint version any time the response format changes in a backwards compatible way.

We do: https://github.com/mastodon/fediverse_auxiliary_service_provider_specifications/blob/main/general/v0.1/provider_specifications.md#versioning

> The `contents` API uses a property called `uri`, the `links` one is called `url`. They should probably be the same.

Admission: I have been aware of the distinction between URIs and URLs for over 25 years now and to this day figuring out when to use which can still make my head spin.

I had a feeling someone would jump on this here, so I tried (and very probably failed) to be precise: Content refers to ActivityPub objects. The AP/AS standards clearly refer to their identifier as an URI. So I used this here.

Links refer to resources that can actually be retrieved from the given location, so while technically also being URIs they are also more specifically URLs.

I also found the distinction nice to see at a glance, that links do *not* refer to AP objects, but are just that, "random web links".

Sorry if I got that wrong ... again :grimacing: 

> Is there any reason not to also support an API call that allows a client to request data about all three types in a single call? It's more efficient than making three API calls, parsing three responses, etc.
> 
> `GET /trends/v0?kind=links,hashtags` for example, to get a response with both the `links` and `hashtags` properties?

The reason is that this is harder to implement and that only recently we had problems with a different, but comparable situation in Mastodon. So I decided to not have this in the first draft and will not have this in the initial implementation (for which we have a tight deadline).

I would not rule this out for a 1.0 of the spec, though, if more people would find it useful.

> It uses a different API endpoint, but Mastodon also has the concept of a trending account. It's part of the suggestions API 

And we plan to also have an `account_recommendation` capability for discovery FASP. So we postponed this until we flesh this out further.

> There are other criteria you might want to use when filtering trending posts. For example, allow the client to fetch a list of trending posts (content) limited to just posts that have video attachments to find trending videos. Or only image attachments. Or polls.

I had not thought of this, but I love this suggestion. Thank you so much!

> Also provide finer control over what is considered to be trending. The document says a trending post might be trending because of replies, favourites, or boosts. But a user might be specifically interested in "Which posts are most favourited" or "Which threads are most active", so allow them to filter by those criteria.

This sounds reasonable as well. As I said, this is currently modeled in a way so we can replicate the different views we have for it in Mastodon. I can see how more fine-grained data / query options could be interesting, but I would really like to postpone this for now. Of course, if other implementers say they need this to replicate how they currently display trends, then I am happy to assign this a higher priority.

nikclayton commented:
## Hashtags

> I hate to inform you, but Mastodon already does this.

I know. And it causes real-world problems like https://github.com/mastodon/mastodon/issues/26518. Not only is that an issue for Mastodon clients, the bottom of that issue has linked issues where it causes interoperability issues with non-Mastodon servers.

Please don't replicate that mistake and embed it further in the ecosystem, for all the reasons I described.

If a Mastodon server wants to casefold the trends it receives from a FASP server to replicate its current broken behaviour, fine. But other Fediverse software should not be forced to the same lowest-common-denominator behaviour.

> One of our goals here is to minimize the risk of all the top results being a variation of the same thing.

I don't think that's attainable in the way you think it is.

Ignore hashtags for a moment, and consider trending links.

Right now my "trending links" timeline shows articles about the Mastodon CEO shakeup from:

- The Verge
- ArsTechnica
- TechCrunch

as well as links to two different articles about Instagram blocking mentions of Pixelfed.

Those are all variations of the same thing. Do you think those links should be folded in to a single trending uber-link?

I hope the answer to that is "no".

Hashtags should behave the same.

> Aside: There is an argument that a future version of this spec might describe a `topic_links` type, which **does** try and aggregate links to different articles about the same thing in to a single item in the response. Similarly, you could imagine a `topic_hashtags` type that does the same thing for hashtags. But right now, since you can't reliably, safely aggregate the data it's better not to.

## Limits

> Putting a limit on this does have two separate purposes, though:

Both of those purposes are implementation specific though, and don't belong in the spec.

## Language

> I was thinking about what we do in Mastodon, which is basically this here: https://github.com/mastodon/mastodon/blob/4a2813158d16914450b6bc4c00e1b3e7edfa26ba/app/helpers/languages_helper.rb

This is another Mastodon bug that you don't want to replicate.

> This is less of a problem for e.g. German, which is mostly spoken in neighboring countries.

:laughs in Swiss German:

Even if Swiss German wasn't a thing, language != country. You can't use language as a proxy for the country, and use that to decide whether or not the user will be interested in.

The only thing you can use language for is "Will the user be able to read and understand this?"

> I am not sure how to solve that.

Two properties.

`language`, the BCP 47 language tag (https://www.rfc-editor.org/bcp/bcp47.txt) which can identify the language, region, variant, and script. You'll need that level of flexibility when distinguishing e.g. between different Chinese scripts.

- `de` - German
- `zh-Hans` - Chinese, simplified
- `zh-yue-Hant` - Chinese, Cantonese, Traditional script
- `de-AT` - Austrian German
- `az-Latn` - Azerbaijani, Latin script
- `az-Cyrl` - Azerbaijain, Cyrillic script

`country`, the ISO 3166-1 alpha 3 code, https://en.wikipedia.org/wiki/List_of_ISO_3166_country_codes has a list.

This allows the client to request all trends, trends in a particular language irrespective of the country, trends in a particular country irrespective of the language, or a mix of the two.

"Show me trending posts written in English about Switzerland" is a reasonable user request. There might be very few results (possibly none), but it should be a valid request.

> I see you use `language` as a property in related specs, like https://github.com/mastodon/fediverse_auxiliary_service_provider_specifications/blob/main/general/v0.1/provider_info.md. Those should be BCP 47 language tags too.

## Merging rank

It matters because if you don't put it in the spec different implementors will have to figure out how to do it, and probably do it in incompatible ways.

If you're not sure how to do it, or even if it's a reasonable thing to do, then that's a strong hint that the bit of the spec that says that clients can consult multiple FASP servers is probably inappropriate (at this time -- supporting multiple servers might be a future goal).

## Returning missing data

I don't think the server should return 0s or sentinel values. It should return the data it has. It is up to the client how to handle missing data in the results.

Related to this, I think the spec should be explicit and say that a server **MUST NOT** return data with holes in it.

By which I mean -- suppose the server had three hours of downtime starting seven hours ago, and has not backfilled the data yet. So the server has a three hour gap in its data.

The only data the server should return is the last four hours, even if the client requests more.

## Distribution charts

The Mastodon web UI is not the only UI that shows this data. Pachli shows charts with correct data for trends. Tusky does something similar, but has some bugs in the presentation.

These are richer than the simple sparklines the Mastodon UI shows.

<img src="https://github.com/user-attachments/assets/f393de58-d980-4efd-9725-28fbafb1961c" width="45%">

Looking at that, I don't think spec will allow this to be replicated -- by returning a single "rank" you throw away the reason for the ranking (e.g., number of distinct mentions, or number of distinct accounts that use the hashtag).

That's probably a mistake -- you're throwing data away too early in the pipeline.

So the value in the `distribution` map should be an object with at least the following properties:

For hashtags and links:

- `count_accounts` - count of distinct accounts that have posted this hashtag or link
- `count_usage` - count of distinct occurrences of this hashtag or link

Different because the same account might post with the same hashtag multiple times. Posting the same link multiple times is probably less frequent, but still possible (e.g., someone replying to different people by sending them the same link).

For content:

- `count_boosts` - Count of boosts of this post
- `count_favourites` - Count of favourites of this post
- `count_mentions` - Count of posts that mention this post (by including it as a link)
- `count_replies` - Count of posts that are immediate replies to this post
- `count_quotes` - (for future use) Count of posts that quote this post using whatever the quoting mechanism turns out to be.

Those could also be a `count` map with keys `boosts`, `favourites`, etc.

> `boosts` or `reblog`? Mastodon uses `reblog`, everyone else calls them boosts.

This also provides a future way of also recording things like the number of distinct reactions a post has, as and when Mastodon supports reactions (which, of course, other servers already do).

## Sorting

Clients are probably going to have to sort the data anyway, especially if they want to provide a richer experience.

Adding this requirement to the server spec makes things slightly more complicated for servers. 

## Data freshness

> What do you mean exactly? As one can only request trends up to the current point in time, data is always expected to be fresh.

A server implementation might not compute the trends in response to the client's request (there's nothing in the spec that says it has to). It could, for example, have a batch job that runs every 30 minutes to compute the trends data.

So the data from one server might be "fresher" than the data from another. That's useful information for the client.

## Caching

> I do not think we can cache anything at the HTTP level here.

You can. Per that previous example, if a server is periodically generating the trends data, instead of doing it on demand, then that trends data can be cached.

## Versioning

The spec says:

> Any change to a specification that runs the risk of being incompatible to existing implementations MUST increase the major version number.

That's too lax. Because you use JSON, amending the spec to add a new property is -- technically -- a backwards compatible change, because clients that don't expect that property will just ignore it when they process the response (generally).

So you might think you don't need to bump the API version.

However, this complicates things for clients, because a client that can handle the new property (if it's present) has to deal with an API that is now returning an optional value.

For an example of this sort of problem, look at https://docs.joinmastodon.org/entities/Notification/. Properties like `report`, `event`, and `moderation_warning` were all added after the API was initially defined, but the API path was never changed from /api/v1/notifications.

Speaking as a client author, this makes my code more complicated, because I have to "sniff" the version of the response from the available properties.

Please don't repeat this mistake -- make it explicit that **adding** properties to the response is not a backwards compatible change, and guarantee you will bump the version number.

## uri and url

> I had a feeling someone would jump on this here, so I tried (and very probably failed) to be precise: Content refers to ActivityPub objects. The AP/AS standards clearly refer to their identifier as an URI. So I used this here.

If what the spec proposal currently calls a `uri` is actually an ActivityPub ID then, for consistency with https://www.w3.org/TR/activitypub/#obj-id, calling it `id` may be preferable.

That AP spec also says that the ActivityPub IDs **MUST** be "Publicly dereferencable URIs, such as HTTPS URIs, with their authority belonging to that of their originating server. (Publicly facing content SHOULD use HTTPS URIs)."

At that point they're definitely URLs, because they contain the information about how to get them (e.g., protocol, server, etc).

So either `url` or `id` would be better than `uri`. `id` is better if you want to hammer home the point that these **must** be AP object IDs, `url` is better if you want to be consistent with `link`.

## Trending accounts

> And we plan to also have an account_recommendation capability for discovery FASP. So we postponed this until we flesh this out further.

That's orthogonal though. Recommending an account is not the same thing as "Here are a list of accounts that are trending".

You might not want to include it in this spec right now, but `/trends/v0/accounts` (hypothetical) and `/discover/v0/accounts` (hypothetical) should be two different things.

Again, just because the Mastodon API put trending accounts under a different part of the namespace doesn't mean you have to repeat the same mistake.

# Other stuff

Additional things that have occurred to me.

## Blocking things from trending

Mastodon allows moderators to block tags (and maybe links and posts, I'm not sure) from the trending view.

Do you intend this functionality to also be available in the FASP server? Or is the idea that it should always provide the unfiltered trends, and it is up to client (the Mastodon or other server that is consuming this data) to implement any blocks.

If the client is filtering the results to apply local blocks then the client is going to sometimes show fewer results than the FASP server provides.

Alternatively, the client could tell the FASP server to apply the block before calculating the trends (`GET /trends/v0/hashtags?block=foo,bar,baz` ?).

I'm mentioning this for completeness -- I think it's a bad idea, because it makes it harder for the FASP server to cache the results (not impossible, just harder), and complicates the code. I think filtering the results should happen downstream of the FASP server.

## Versioning

The examples in the doc use `v0` in the path. In https://github.com/mastodon/fediverse_auxiliary_service_provider_specifications/blob/main/general/v0.1/provider_info.md versions are shown as examples like `1.0`.

It doesn't look like you're using Semantic Versioning.

Please use Semantic Versioning.

I think https://github.com/mastodon/fediverse_auxiliary_service_provider_specifications/blob/main/general/v0.1/provider_specifications.md#http-api-endpoints saying that only the major API version is in the API path is a mistake. The API path should include the full version to avoid any ambiguity.

## Provide a `capabilities` escape hatch

There's no mechanism for a FASP server to expose sets of capabilities other than the official ones. This will make experimentation in this space difficult and lead to interoperability problems.

Please steal the "unique identifier" example from https://codeberg.org/fediverse/fep/src/branch/main/fep/9fde/fep-9fde.md#a-unique-identifier-for-each-operation.

So instead of (as documented) ids of `trends` and `account_search` call them something like:

- `org.fedidiscovery.api.trends`
- `org.fedidiscovery.api.account_search`

Now, if someone wants to run a FASP server that provides a new capability that hasn't been standardised yet they can call it something like `tld.my-domain.api.my-new-capability` and be sure it won't clash with anything in the future. If/when it becomes standardised it can migrate to an `org.fedidiscovery.api....` name.

## Privacy

https://github.com/mastodon/fediverse_auxiliary_service_provider_specifications/blob/main/general/v0.1/provider_specifications.md#http-api-endpoints says

> Since FASP specifications define exactly what data the FASP can receive, they MUST include a summary noting which of this data is to be considered sensitive and why it needs to be processed by the FASP.

The trends specification does not include this mandatory section.

astro commented on discovery/trends/v0.1/trends.md:
```diff
@@ -0,0 +1,355 @@
+  server uses several FASP to query for trends that all run the same
+  software, it MUST be possible to merge results according to rank to
+  get the correct order.
+* `distribution`: An object with one key per hour requested. The keys
```
This sequence number-indexed object could be a simple array.

oneiros replied:
Good point! Thanks a lot. I changed it accordingly.

That being said, I might still remove this altogether as I am no longer convinced of its usefulness.

astro replied:
I vote for keeping: I think these numbers are very useful for Mastodon's sparklines and the graphs in https://fedi.buzz

oneiros replied:
I am not against reintroducing something like this at a later stage, but having thought about it some more and having tried to integrate this into Mastodon, I do not think that in its current state it makes a lot of sense:

1. Mastodon's API returns specific metrics for the last 7 days (if available), while this draft would only give one (incompatible) metric for the requested number of hours. It is not compatible / compararable. So this would not help Mastodon and indeed we opted for not using it at all in the first implementation.
2. A trend is often defined as an uptick in activity compared to a similar timeframe in the past. These numbers would not reflect that, making me wonder if displaying them would even make sense. I could imagine returning more data points here to make it more relevant, but how much more? I am not a fan of hardcoding this to 7 days as Mastodon does (or any other fixed value for that matter).
3. I no longer believe the normalization requirement here is sensible. I would probably opt for absolute numbers instead.
4. The discussion here and also some experience from Mastodon shows that people sometimes want to make sense of these numbers and also expect a direct and obvious relationship to the ranking of trends. But this is just not possible in Mastodon and imo also hard to achieve here. I would rather not include any numbers at all than include numbers that lead to confusion (or even complaints).

Again, I am totally open to re-introduce a better version of this. If you have any ideas for that, please open an issue. But for now I will remove this.

astro commented on discovery/trends/v0.1/trends.md:
```diff
@@ -0,0 +1,355 @@
+All responses include a list of results. Individual result objects
+share the following common keys and values:
+
+* `rank`: A positive integer less than or equal `100` representing the
```
A *relative rank* will make it harder to compare trends over time.

oneiros replied:
Yes! Which is intentional. This is API is only about _current_ trends. Show the user a couple of things are relevant _right now_.

I know that one can easily construct other use cases. Being able to query and compare historic trends for example is probably super interesting, especially for researchers. However, this does not fall under the umbrella of "search and discovery", so is totally out-of-scope here.

astro replied:
Even with *current* trends, relative ranks are useless if anyone wants to combine the data from multiple FASPs.

oneiros replied:
Not necessarily. I addressed this in this comment here: https://github.com/mastodon/fediverse_auxiliary_service_provider_specifications/pull/45#issuecomment-2650353942

astro commented on discovery/trends/v0.1/trends.md:
```diff
@@ -0,0 +1,355 @@
+requests that allow - with one exception - using the same optional
+parameters:
+
+* `withinLastHours`: This MUST be a positive integer specifying the
```
How is can this be rounded to actually available time periods? For example, my FASP holds trends for 4-hour, 1-day, and 1-week intervals in precomputed materialized views.

Consequently, the actual time period should be indicated in the response.

oneiros replied:
Good question. I am not sure if this fine-grained option to pass an arbitrary number of hours is even necessary. Maybe we could come up with some pre-defined timeframes (e.g. last hour, last 6, 12, 24, 48 and one week).

My only goal here was to offer something that is more flexible than what we currently have in Mastodon.

I would love to hear other people's thought on this.

ThisIsMissEm replied:
Maybe a Trend FASP could advertise which granularity and retention periods it supports?

astro commented on discovery/trends/v0.1/trends.md:
```diff
@@ -0,0 +1,355 @@
+Specification identifier: `trends`
+
+The `trends` capability allows fediverse servers to query FASP for
+content, hashtags and links that are currently trending.
```
Does an FASP always have to provide all three capabilities at the same time? Would it be possible to offer just trending hashtags?

oneiros replied:
Interesting idea. I do not currently think we should split this up any further, but we definitely could in the future if there is demand.

nikclayton commented:
Ping. Is this work paused?

oneiros commented:
> ## Hashtags
> [...] 
> Please don't replicate that mistake and embed it further in the ecosystem, for all the reasons I described.

I am afraid we will have to agree to disagree here. We discussed this again internally and we really care about all variations of popular hashtags (e.g. `#catsofmastodon`) to be treated as being the same.

That being said, the paragraph in my draft about this only says:

> FASP SHOULD consider normalization to minimize the risk of obvious duplicates

So there is nothing preventing anyone from building a more "correct" trends provider.

> ## Limits
> 
> > Putting a limit on this does have two separate purposes, though:
> 
> Both of those purposes are implementation specific though, and don't belong in the spec.

Personally, I still firmly believe that putting such a limit in the specification would be beneficial to make it absolutely clear that no copious amount of data should be kept.

But you were not the only one having doubts about this, so I changed the language here to make at least one week of data mandatory, but allow for more optionally.

> ## Language

I changed the spec to use language tags. After all this is what ActivityPub uses anyway. Thanks for pointing this out.

> ## Merging rank
> 
> It matters because if you don't put it in the spec different implementors will have to figure out how to do it, and probably do it in incompatible ways.
> 

But that is the point. If we allow different provider software to calculate trends in vastly different ways, I do not know how we can define an objective, comparable rank that works for everyone. (If you do, please let me know!)

That is why the spec says that this has to only be comparable between providers *that run the exact same software*. 

Fediverse servers can decide to:

* Only ever get trends from a single provider
* Only merge trends from several providers running the same software, possibly displaying different trend sources in different ways (tabs, columns, whatever)
* Merge trends from different sources according to rank anyway even though that is not strictly correct, because it might still fulfill the end goal here

I can easily imagine that this is an area for future improvement once there are competing provider implementations and integrations in different fediverse servers. But right now I am not sure how we could do any better.

> ## Returning missing data

A good point that I will have to ponder more.

> ## Distribution charts

Thank you for sharing your perspective. Apart from what you mentioned, I learned there are other issues here as well.

I am currently leaning towards removing the distribution array altogether. In its current form it does not make much sense I fear.

Maybe something to re-introduce at a later point.

> ## Data freshness
> 
> A server implementation might not compute the trends in response to the client's request (there's nothing in the spec that says it has to). It could, for example, have a batch job that runs every 30 minutes to compute the trends data.
> 
> So the data from one server might be "fresher" than the data from another. That's useful information for the client.

Good point, but imo something for a later version, if we see demand for this.

> ## Caching
> 
> > I do not think we can cache anything at the HTTP level here.
> 
> You can.

We cannot because of the HTTP Message Signatures we use. Caching makes a lot of sense here on different levels. HTTP is not one them though.

> ## uri and url

Thanks. I see your point. Still have not made up my mind about this.

> ## Blocking things from trending
> 
> Mastodon allows moderators to block tags (and maybe links and posts, I'm not sure) from the trending view.
> 
> Do you intend this functionality to also be available in the FASP server? Or is the idea that it should always provide the unfiltered trends, and it is up to client (the Mastodon or other server that is consuming this data) to implement any blocks.
> 
> If the client is filtering the results to apply local blocks then the client is going to sometimes show fewer results than the FASP server provides.
> 
> Alternatively, the client could tell the FASP server to apply the block before calculating the trends (`GET /trends/v0/hashtags?block=foo,bar,baz` ?).
> 
> I'm mentioning this for completeness -- I think it's a bad idea, because it makes it harder for the FASP server to cache the results (not impossible, just harder), and complicates the code. I think filtering the results should happen downstream of the FASP server.

I agree, at least for now. Let us see if and how there is demand for this once other fediverse server implementers try to integrate this.

> ## Privacy
> 
> https://github.com/mastodon/fediverse_auxiliary_service_provider_specifications/blob/main/general/v0.1/provider_specifications.md#http-api-endpoints says
> 
> > Since FASP specifications define exactly what data the FASP can receive, they MUST include a summary noting which of this data is to be considered sensitive and why it needs to be processed by the FASP.
> 
> The trends specification does not include this mandatory section.

That is correct. I am still waiting on how this API will turn out before committing on something for this section. Right now it would be very much empty, as the current draft does not allow the fediverse server to send any PII to the FASP (as part of the trends API!).

oneiros commented:
Thanks to everyone who already chimed in here. I was able to make some improvements based on your comments and I am sure we can iron out more problems in the future.

As you might be aware, we are currently working towards meeting some deadlines set by our NGI Search project. Given that I will merge this PR in the coming days, even if maybe we have not found a satisfying answer to all open questions.

This does not mean that I/we do not take them seriously. I will try to open issues for the most urgent ones, so we can discuss them further. If you think I missed something important or if you have further ideas or questions, feel free to open issues as well.

We will not declare any of this as final (or `1.0`) any time soon, so there will be time for improvement. But I will be busy with other parts of the project in the coming weeks, so please bear with me if I do not react in a timely fashion.

nikclayton commented:
# Hashtags

This decision to throw users with accessibility needs under the bus is very disappointing.

Again -- you don't need to do this at the FASP server level. The server can provide un-case-folded trends to the client, and the client can decide how to merge them.

Weasel wording around "should consider normalisation" at the FASP server level will just cause more interoperability problems down the line.

# Limits

Still shouldn't be in the spec as a minimum thing the server has to provide.

Consider a trends server, for example, that is computing minute-by-minute trends updates over the last hour, vs. a server that is computing hourly updates.

A client might connect to both of them to provide short term and long term trends. There's no need for the minute-to-minute server to produce a week's worth of results.

# Merging rank

If this isn't going to work then be explicit about in the spec, and say that clients should not try and merge data from different FASP servers.

If you expect that merging rank is only possible between providers *that run the exact same software* then the spec would need to describe how clients can determine the (software, version) tuple that produced the ranking.

> More accurately, it's not the (software, version) tuple, it's whatever algorithm is being used to calculate the rank.

But that's still probably a bad idea, because now the client has to handle the case where it was connected to two FASP servers that were running the same (software, version), but one of them has upgraded to a newer version.

What is the client supposed to do in this case?

# Distribution charts

If you remove this then isn't this going to remove functionality clients are currently providing?

# Data freshness

I see a need for this now. Clients can expose the data freshness (even better, when trends are expected to update) as information to users, so they don't keep refreshing the Trends page to see if there's new content.

# Caching

> We cannot because of the HTTP Message Signatures we use. Caching makes a lot of sense here on different levels. HTTP is not one them though.

I do not understand this. Are you suggesting that clients should not be able to cache the results of making a `GET /trends/v0/hashtags` request to a FASP server?
